libs/esp32spi/adafruit_esp32spi.ts

```diff
@@ -0,0 +1,704 @@
+    export const SOCKET_LISTEN = 1
+    export const SOCKET_SYN_SENT = 2
+    export const SOCKET_SYN_RCVD = 3
+    export const SOCKET_ESTABLISHED = 4
+    export const SOCKET_FIN_WAIT_1 = 5
+    export const SOCKET_FIN_WAIT_2 = 6
+    export const SOCKET_CLOSE_WAIT = 7
+    export const SOCKET_CLOSING = 8
+    export const SOCKET_LAST_ACK = 9
+    export const SOCKET_TIME_WAIT = 10
+    export const WL_NO_SHIELD = 0xFF
+    export const WL_NO_MODULE = 0xFF
+    export const WL_IDLE_STATUS = 0
+    export const WL_NO_SSID_AVAIL = 1
+    export const WL_SCAN_COMPLETED = 2
+    export const WL_CONNECTED = 3
+    export const WL_CONNECT_FAILED = 4
+    export const WL_CONNECTION_LOST = 5
+    export const WL_DISCONNECTED = 6
+    export const WL_AP_LISTENING = 7
+    export const WL_AP_CONNECTED = 8
+    export const WL_AP_FAILED = 9
+
+    export class ESP_SPIcontrol {
+        _spi: SPI;
+        _debug: number;
+        _gpio0: DigitalInOutPin;
+        _cs: DigitalInOutPin;
+        _ready: DigitalInOutPin;
+        _reset: DigitalInOutPin;
+        _pbuf: Buffer;
+        _socknum_ll: any; /** TODO: type **/
+
+        static instance: ESP_SPIcontrol;
+
+        static TCP_MODE = 0
+        static UDP_MODE = 1
+        static TLS_MODE = 2
+
+        constructor() {
+            this._pbuf = control.createBuffer(1);
+        }
+
+        private log(priority: number, msg: string) {
+            console.log(msg);
+        }
+
+        public reset(): void {
+            /** Hard reset the ESP32 using the reset pin */
+            if (this._debug) {
+                print("Reset ESP32")
+            }
+
+            if (this._gpio0)
+                this._gpio0.digitalWrite(true);
+            this._cs.digitalWrite(true)
+            this._reset.digitalWrite(false)
+            // reset
+            pause(10)
+            this._reset.digitalWrite(true)
+            // wait for it to boot up
+            pause(750)
+            if (this._gpio0)
+                this._gpio0.digitalRead();
+        }
+
+
+        // we're ready!
+        // pylint: disable=too-many-branches
+
+        // header + end byte
+        // parameter
+        // size byte
+        // 2 of em here!
+        // we may need more space
+        // handle parameters here
+        // %d is %d bytes long" % (i, len(param)))
+        // wait up to 1000ms
+        // ok ready to send!
+        // pylint: disable=no-member
+        // pylint: disable=too-many-branches
+        private _read_byte(): number {
+            this._spi.transfer(undefined, this._pbuf); // TODO
+            if (this._debug >= 3) {
+                print(`\t\tRead: ${this._pbuf[0]}`);
+            }
+            return this._pbuf[0]
+        }
+
+        /*
+        private _read_bytes(spi: SPI, buffer: Buffer, start: number = 0, end: number = null): void {
+            if (!end) {
+                end = buffer.length
+            }
+
+            spi.transfer(undefined, buffer); // TODO
```

Comment:
we don't support range for reading into a buffer.

Reply:
this doesn't seem to be used